File: app/src/main/java/com/lin/read/adapter/ScanTypeAdapter.java
package com.lin.read.adapter;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.CheckBox;
import android.widget.CompoundButton;
import android.widget.TextView;

import com.lin.read.R;
import com.lin.read.filter.scan.ScanTypeInfo;

import java.util.List;

/**
 * Created by lisonglin on 2017/10/12.
 */

public class ScanTypeAdapter extends RecyclerView.Adapter<ScanTypeAdapter.ViewHolder> {
    private Context context;
    private List<ScanTypeInfo> data;

    private boolean isHold = false;
    private int layoutId;

    public ScanTypeAdapter(Context context, List<ScanTypeInfo> data) {
        this.context = context;
        this.data = data;
    }

    public ScanTypeAdapter(Context context,List<ScanTypeInfo> data, int layoutId){
        this.context = context;
        this.data = data;
        this.layoutId = layoutId;
    }

    @Override
    public ViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        View view = LayoutInflater.from(context).inflate(layoutId == 0 ? R.layout.item_scan_type : layoutId, null);
        ViewHolder holder = new ViewHolder(view);
        return holder;
    }

    @Override
    public void onBindViewHolder(ViewHolder holder, int position) {
        isHold = true;
        final ScanTypeInfo typeInfo = data.get(position);
        holder.text.setText(typeInfo.getText());
        holder.text.setTag(position);
        holder.checkBox.setChecked(typeInfo.isChecked());
        isHold = false;
    }

    @Override
    public int getItemCount() {
        return data == null ? 0 : data.size();
    }

    class ViewHolder extends RecyclerView.ViewHolder {

        private TextView text;
        private CheckBox checkBox;

        public ViewHolder(View itemView) {
            super(itemView);
            text = (TextView) itemView.findViewById(R.id.scan_type_text);
            checkBox = (CheckBox) itemView.findViewById(R.id.scan_type_checkbox);

            checkBox.setOnTouchListener(new View.OnTouchListener() {
                @Override
                public boolean onTouch(View v, MotionEvent event) {
                    if (checkBox.isChecked()) {
                        return true;
                    }
                    return false;
                }
            });

            checkBox.setOnCheckedChangeListener(new CompoundButton.OnCheckedChangeListener() {
                @Override
                public void onCheckedChanged(CompoundButton buttonView, boolean isChecked) {
                    int position = (int) text.getTag();
                    if (!isHold) {
                        if(isChecked){
                            if(onScanItemClickListener !=null){
                                onScanItemClickListener.onItemClick(position,text.getText().toString());
                            }
                        }
                        setCheckedStatus(text.getText().toString());
                    }
                }
            });
        }
    }

    private void setCheckedStatus(String text) {
        if (data == null || data.size() == 0 || TextUtils.isEmpty(text)) {
            return;
        }
        for (ScanTypeInfo item : data) {
            if (text.equals(item.getText())) {
                item.setChecked(true);
            } else {
                item.setChecked(false);
            }
        }
        notifyDataSetChanged();
    }

    public ScanTypeInfo getCheckedInfo() {
        if (data == null || data.size() == 0) {
            return null;
        }
        for (ScanTypeInfo item : data) {
            if (item.isChecked()) {
                return item;
            }
        }
        return null;
    }

    public String getCheckedText() {
        if (data == null || data.size() == 0) {
            return null;
        }
        for (ScanTypeInfo item : data) {
            if (item.isChecked()) {
                return item.getText();
            }
        }
        return null;
    }

    public int getCheckedPosition() {
        if (data == null || data.size() == 0) {
            return -1;
        }
        for (int i = 0; i < data.size(); i++) {
            if (data.get(i).isChecked()) {
                return i;
            }
        }
        return -1;
    }

    public interface OnScanItemClickListener {
        void onItemClick(int position,String clickText);
    }

    private OnScanItemClickListener onScanItemClickListener;

    public void setOnScanItemClickListener(OnScanItemClickListener onScanItemClickListener) {
        this.onScanItemClickListener = onScanItemClickListener;
    }
}

File: app/src/main/java/com/lin/read/adapter/ScanTypeAdapter.kt
package com.lin.read.adapter

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.CheckBox
import android.widget.TextView
import com.lin.read.R
import com.lin.read.filter.scan.ReadScanTypeData
import com.lin.read.filter.scan.VarPair

class ScanTypeAdapter @JvmOverloads constructor(private val context: Context, private val data: List<VarPair<ReadScanTypeData,Boolean>>, use4Words: Boolean = false) : RecyclerView.Adapter<ScanTypeAdapter.ViewHolder>() {
    private var isBinding = false
    private var layoutId: Int
    var onScanItemClickListener: ((Int, String) -> Unit)? = null
    val onScanItemClickListenerMap: HashMap<String, (Int, String) -> Unit>

    init {
        layoutId = if (use4Words) R.layout.item_scan_type_4_chars else R.layout.item_scan_type
        onScanItemClickListenerMap = hashMapOf()
        data.forEach { it.second = it.first.default }
    }
    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): ViewHolder {
        return ViewHolder(LayoutInflater.from(context).inflate(layoutId, null))
    }

    override fun getItemCount(): Int {
        return data.size
    }

    override fun onBindViewHolder(holder: ViewHolder?, position: Int) {
        isBinding = true
        holder?.run {
            text.text = data[position].first.name
            text.tag = position
            checkBox.isChecked = data[position].second
        }
        isBinding = false
    }

    fun getCheckedInfo(): ReadScanTypeData = data.first { it.second }.first

    fun getCheckedText(): String = data.first { it.second }.first.name

    private fun setCheckedStatus(text: String) {
        data.forEach { it.second = it.first.name == text }
        notifyDataSetChanged()
    }

    inner class ViewHolder(view: View) : RecyclerView.ViewHolder(view) {
        var text:TextView
        var checkBox: CheckBox
        init {
            text = view.findViewById(R.id.scan_type_text) as TextView
            checkBox = (view.findViewById(R.id.scan_type_checkbox) as CheckBox).apply {
                setOnTouchListener { _, _ -> isChecked }
                setOnCheckedChangeListener { _, isChecked ->
                    if (!isBinding) {
                        setCheckedStatus(this@ViewHolder.text.text.toString())
                        if (isChecked) {
                            onScanItemClickListener?.invoke(this@ViewHolder.text.tag as Int, this@ViewHolder.text.text.toString())
                            onScanItemClickListenerMap[this@ViewHolder.text.text.toString()]?.invoke(this@ViewHolder.text.tag as Int, this@ViewHolder.text.text.toString())
                        }
                    }
                }
            }
        }
    }
}